src/main/java/core/basesyntax/ArrayList.java

```diff
@@ -1,48 +1,142 @@
 package core.basesyntax;
 
+import java.util.NoSuchElementException;
+import java.util.Objects;
+
 public class ArrayList<T> implements List<T> {
+    private static final int DEFAULT_CAPACITY = 10;
+    private static final double GROWTH_RATE = 1.5;
+    private int size;
+    private int newCapacity;
+    private Object[] data;
+
+    public ArrayList() {
+        data = new Object[DEFAULT_CAPACITY];
+        size = 0;
+        newCapacity = DEFAULT_CAPACITY;
+    }
+
+    private void setData(T value, int index) {
+        if (isNeedGrow()) {
+            grow(this.data.length);
+        }
+        data[index] = value;
+        size++;
+    }
+
+    private void checkIndex(int index) {
+        if (index > size - 1 || index < 0) { //
+            throw new ArrayListIndexOutOfBoundsException("Out of bounds exception");
+        }
+    }
+
+    private boolean isNeedGrow() {
+        return size == newCapacity;
+    }
+
+    private Object[] trimToSize(Object[] objects) {
+        System.arraycopy(objects, 0, data, 0, size);
+        return objects;
+    }
+
+    private void grow(int capacity) {
+        newCapacity = (int) ((capacity) * GROWTH_RATE);
+        Object[] data = new Object[newCapacity];
+        System.arraycopy(this.data, 0, data, 0, this.data.length);
+        this.data = data;
+    }
+
+    private void copyDataWithDecreaseLength(int index) {
+        Object[] data = new Object[this.data.length];
+        System.arraycopy(this.data, 0, data, 0, index);
+        System.arraycopy(this.data, index + 1, data, index, size - 1 - index);
```

Comment:
```suggestion
        Object[] data = this.data;
        System.arraycopy(this.data, index + 1, data, index, size - 1 - index);
```
те саме можна зробити в методі copyDataWithIncreaseLength(int index)